share/smack/lib/string.c

```diff
@@ -10,12 +10,88 @@ char *strcpy(char *dest, const char *src) {
   return save;
 }
 
+char *strncpy(char *dest, const char *src, size_t n) {
+  size_t i;
+
+  for (i = 0; i < n && src[i] != '\0'; i++)
+    dest[i] = src[i];
+  for ( ; i < n; i++)
+    dest[i] = '\0';
+
+  return dest;
+}
+
 size_t strlen(const char *str) {
   size_t count = 0;
-  while (str[count] != 0) count++;
+  while (str[count]) count++;
   return count;
 }
 
+// comparison
+
+int strcmp(const char *s1, const char *s2) {
+  size_t n;
+  for (n = 0; s1[n] == s2[n]; n++)
+    if (s1[n] == '\0')
+      return 0;
+  return s1[n] - s2[n];
+}
+
+int strncmp(const char *s1, const char *s2, size_t n) {
+ 
+// what do we do if n is bigger than s1 and s2? 
```

Comment:
It is fine to remove this comment. This is a known problem in real-world implementations.